AST-based operations

klavs opened the pull request:
Allow creating operations out of AST.
This is a non-breaking transitional to enable AST-based links, local state resolvers.

This change add as slight performance reduction because documents must be printed to string before sending them over the network.
The remedy for that is building AST at build-time. Currently that is possible using `package:gql_code_gen`.

klavs commented:
@HofmannZ, here's the non-breaking change I was referring to in https://github.com/zino-app/graphql-flutter/pull/414

HofmannZ commented on packages/graphql/lib/src/core/query_options.dart:
```diff
@@ -67,12 +69,17 @@ class Policies {
 class BaseOptions extends RawOperationData {
   BaseOptions({
-    @required String document,
+    @Deprecated('Use documentNode instead') String document,
```
Good to first deprecated support for `document`. Maybe rephrase the message to:

```
'The "document" option has been deprecated, use "documentNode" instead'
```

klavs replied:
OK, added a more verbose deprecation message

HofmannZ commented on packages/graphql/lib/src/core/query_options.dart:
```diff
@@ -194,12 +207,25 @@ typedef dynamic UpdateQuery(
+            (document == null && documentNode != null)),
+        assert(updateQuery != null);
+
+  DocumentNode documentNode;
```
Why not use getters and setters?

klavs replied:
Added the missing initialisation to the constructor

HofmannZ commented on packages/graphql/lib/src/core/raw_operation_data.dart:
```diff
@@ -31,7 +45,7 @@ class RawOperationData {
     // XXX there is a bug in the `graphql_parser` package, where this result might be
     // null event though the operation name is present in the document
-    _operationName ??= getOperationName(document);
+    _operationName ??= getLastOperationName(documentNode);
```
Can we test if this bug is fixed?

klavs replied:
Removing the comment. It's not too helpful in describing the exact conditions of the bug.

If any bug in parser appears, we'll file a bug to `package:gql` and we'll fix it there.

klavs commented:
@HofmannZ I addressed the feedback and signed the commits.

I think the CI fails incorrectly – we should be able to deprecate our own API and use it while it is deprecated.

cc: @mainawycliffe

klavs commented:
added info about AST documents to the readme.

Note that I'm assuming the version `2.2.0`

lookfirst commented:
This is awesome!

HofmannZ commented:
:tada: This PR is included in version 2.1.1-beta.3 :tada:

The release is available on [GitHub release](https://github.com/zino-app/graphql-flutter/releases/tag/v2.1.1-beta.3%40beta)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

HofmannZ commented:
:tada: This PR is included in version 3.0.0-beta.1 :tada:

The release is available on [GitHub release](https://github.com/zino-app/graphql-flutter/releases/tag/v3.0.0-beta.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

HofmannZ commented:
:tada: This PR is included in version 3.0.0 :tada:

The release is available on [GitHub release](https://github.com/zino-app/graphql-flutter/releases/tag/v3.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
